lms/static/scripts/frontend_apps/utils/onedrive-picker-client.js

```diff
@@ -0,0 +1,70 @@
+import { loadOneDriveAPI } from './onedrive-api-client';
+import { PickerCanceledError } from './google-picker-client';
+
+export class OneDrivePickerClient {
+  /**
+   * @param {object} options
+   * @param {string} options.clientId
+   * @param {string} options.redirectURI - the URL that is used to launch the picker
+   */
+  constructor({ clientId, redirectURI }) {
+    loadOneDriveAPI();
+
+    this._filePickerOptions = {
+      clientId,
+      action: 'share',
+      multiSelect: false,
+      viewType: 'files', // The type of item that can be selected.
+      advanced: {
+        redirectUri: redirectURI,
+        filter: '.pdf',
+        createLinkParameters: { type: 'view', scope: 'anonymous' },
+      },
+    };
+  }
+
+  /**
+   * Opens the OneDrive picker
+   *
+   * @returns Promise<{url: string}> - the URL of the file
+   * @throws {Error} - there are two types of errors:
+   *   - PickerCancelledError: raise when the user cancels the OneDrive dialog picker
+   *   - Other errors: raise when (1) the OneDrive client fails to load (`window.OneDrive` is
+   *       undefined), or when (2) the picker fails in a different way.
+   */
+  async showPicker() {
+    return new Promise((resolve, reject) => {
+      const success = (/** @type {any} */ file) => {
+        const sharingURL = file.value[0].permissions[0].link.webUrl;
+        const url = OneDrivePickerClient.encodeSharingURL(sharingURL);
+        resolve({ url });
+      };
+      const cancel = () => reject(new PickerCanceledError());
+      const error = (/** @type {Error} */ error) => reject(error);
+      window.OneDrive.open({
```

Comment:
This call could fail if the OneDrive script has not completed loading before `showPicker` is called. Also I think it makes sense to consider the `window.OneDrive` global an implementation detail of the `loadOneDriveAPI` helper.

What I would suggest to do instead is to save the promise from `loadOneDriveAPI` in the constructor and then await the result here before calling the `open` method on it.